service/history/timerBuilder.go

```diff
@@ -222,10 +222,26 @@ func (tb *timerBuilder) GetUserTimer(timerID string) (bool, *persistence.TimerIn
 }
 
 // IsTimerExpired - Whether a timer is expired w.r.t reference time.
-func (tb *timerBuilder) IsTimerExpired(td *timerDetails, referenceTime time.Time) bool {
+func (tb *timerBuilder) IsTimerExpired(td *timerDetails, reference time.Time) bool {
+	// TODO revisit this timer resolution issue #2102
+
 	// Cql timestamp is in milli sec resolution, here we do the check in terms of second resolution.
-	expiry := td.TimerSequenceID.VisibilityTimestamp.Unix()
-	return expiry <= referenceTime.Unix()
+	expireTime := td.TimerSequenceID.VisibilityTimestamp.UnixNano()
+	expireTimeSeconds := expireTime / 1e9
+	expireTimeNanoSeconds := expireTime % 1e9
+
+	referenceTime := reference.UnixNano()
+	referenceTimeSeconds := referenceTime / 1e9
+	referenceTimeNanoSeconds := referenceTime % 1e9
+
+	// do a tiny sleep here to guarantee timer fire after schedule time + timeout time
+	if expireTimeSeconds <= referenceTimeSeconds {
+		if expireTimeNanoSeconds > referenceTimeNanoSeconds {
+			time.Sleep(time.Nanosecond * time.Duration(expireTimeNanoSeconds-referenceTimeNanoSeconds))
```

Comment:
Can you check what the recommended minimum for `time.Sleep` is? Super short sleeps are not always handled well..

Reply:
Somewhat related, but take a look at this - https://github.com/golang/go/issues/17696.. Windows minimum timer duration is 1ms or 15.6ms.. I am not sure what Linux exposes to users.. I remember reading for super short nanosecond sleeps its better to do (cannot find reference now).. 

```
while time.Now() < expTime {
    yield()
}
```

Reply:
did 2 tests, one on mac, one on linux (4.something)
time interval on sleep 1 nano second produces double digits micro seconds interval.

just keep this PR open, in case java side needs is.